File: python/check_if_a_string_can_break_another_string.py
"""
#slower solution
class Solution:
    def checkIfCanBreak(self, s1: str, s2: str) -> bool:
        ascii_values1 = [0 for i in range(26)]
        ascii_values2 = [0 for i in range(26)]
        for ch in s1:
            ascii_values1[ord(ch) - 97] += 1
        for ch in s2:
            ascii_values2[ord(ch) - 97] += 1
        nums1, nums2 = [], []
        for i in range(len(ascii_values1)):
            for j in range(ascii_values1[i]):
                nums1.append(i)
            for j in range(ascii_values2[i]):
                nums2.append(i)
        s1_breaks_s2, s2_breaks_s1 = True, True
        for i in range(len(nums1)):
            if nums1[i] > nums2[i]:
                s1_breaks_s2 = False
            if nums2[i] > nums1[i]:
                s2_breaks_s1 = False
            if not s1_breaks_s2 and not s2_breaks_s1:
                return False
        return True
"""
class Solution:
    def checkIfCanBreak(self, s1: str, s2: str) -> bool:
        #sort strings (since we only need to iterate through them, there's no need to join them back into strings)
        ss1, ss2 = sorted(s1), sorted(s2)
        s1_breaks_s2, s2_breaks_s1 = True, True
        #iterate through sorted strings to determine if one breaks the other
        for i in range(len(s1)):
            if ord(ss1[i]) > ord(ss2[i]):
                s2_breaks_s1 = False
            if ord(ss2[i]) > ord(ss1[i]):
                s1_breaks_s2 = False
            if not s1_breaks_s2 and not s2_breaks_s1:
                return False
        return True

File: java/checkIfAStringCanBreakAnotherString.java
class Solution {
    public boolean checkIfCanBreak(String s1, String s2) {
        int l = s1.length();
        //store each string as an array of numbers, corresponding to the ascii values
        int[] nums1 = new int[l];
        int[] nums2 = new int[l];
        for (int i = 0; i < l; i++) {
            nums1[i] = (int) s1.charAt(i);
            nums2[i] = (int) s2.charAt(i);
        }
        //sort the arrays
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        boolean s1BreaksS2 = true;
        boolean s2BreaksS1 = true;
        //compare the arrays one number at a time to determine if one breaks the other 
        for (int i = 0; i < l; i++) {
            if (nums1[i] > nums2[i]) {
                s2BreaksS1 = false;
            }
            if (nums2[i] > nums1[i]) {
                s1BreaksS2 = false;
            }
            if (!s1BreaksS2 && !s2BreaksS1) {
                return false;
            }
        }
        return true;
    }
}